doc/_build/man/nginx-build.1
.\" Man page generated from reStructuredText.
.
.TH "NGINX-BUILD" "1" "December 28, 2014" "0.2.0" "nginx-build"
.SH NAME
nginx-build \- seamless nginx builder
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH DESCRIPTION
.sp
nginx\-build is a seamless nginx builder.
.SH OPTIONS
.sp
\fB\-version\fP
.INDENT 0.0
.INDENT 3.5
show version of nginx\-build
.UNINDENT
.UNINDENT
.sp
\fB\-versions\fP
.INDENT 0.0
.INDENT 3.5
show versions of nginx
.UNINDENT
.UNINDENT
.sp
\fB\-v [nginx\-version]\fP
.INDENT 0.0
.INDENT 3.5
specify nginx version
.UNINDENT
.UNINDENT
.sp
\fB\-j [jobs]\fP
.INDENT 0.0
.INDENT 3.5
jobs to build nginx
.UNINDENT
.UNINDENT
.sp
\fB\-c [configure]\fP
.INDENT 0.0
.INDENT 3.5
specify configuration file for building nginx
.UNINDENT
.UNINDENT
.sp
\fB\-clear\fP
.INDENT 0.0
.INDENT 3.5
remove entries in working directory
.UNINDENT
.UNINDENT
.sp
\fB\-d [working directory]\fP
.INDENT 0.0
.INDENT 3.5
specify working directory for building nginx
.UNINDENT
.UNINDENT
.sp
\fB\-m [modules3rd.ini]\fP
.INDENT 0.0
.INDENT 3.5
configuration file for 3rd party modules
.UNINDENT
.UNINDENT
.sp
\fB\-pcre\fP
.INDENT 0.0
.INDENT 3.5
embed PCRE statically for building nginx
.UNINDENT
.UNINDENT
.sp
\fB\-pcreversion [pcre\-version]\fP
.INDENT 0.0
.INDENT 3.5
specify PCRE version
.UNINDENT
.UNINDENT
.sp
\fB\-zlib\fP
.INDENT 0.0
.INDENT 3.5
embed ZLIB statically for building nginx
.UNINDENT
.UNINDENT
.sp
\fB\-zlibversion [zlib\-version]\fP
.INDENT 0.0
.INDENT 3.5
specify ZLIB version
.UNINDENT
.UNINDENT
.sp
\fB\-openssl\fP
.INDENT 0.0
.INDENT 3.5
embed OpenSSL statically for building nginx
.UNINDENT
.UNINDENT
.sp
\fB\-opensslversion [openssl\-version]\fP
.INDENT 0.0
.INDENT 3.5
specify OpenSSL version
.UNINDENT
.UNINDENT
.sp
\fB\-openresty\fP
.INDENT 0.0
.INDENT 3.5
download openresty instead of nginx
.UNINDENT
.UNINDENT
.sp
\fB\-openrestyversion [openresty\-version]\fP
.INDENT 0.0
.INDENT 3.5
openresty version
.UNINDENT
.UNINDENT
.sp
\fB\-verbose\fP
.INDENT 0.0
.INDENT 3.5
enable verbose mode
.UNINDENT
.UNINDENT
.SH AUTHOR
Tatsuhiko Kubo
.SH COPYRIGHT
2014, Tatsuhiko Kubo
.\" Generated by docutils manpage writer.
.


nginx-build.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"runtime"
	"strings"
	"syscall"
)

func main() {
	var dependencies []StaticLibrary
	parallels := 0
	done := make(chan bool)

	// set parallel numbers
	runtime.GOMAXPROCS(runtime.NumCPU())

	// Parse flags
	version := flag.String("v", NGINX_VERSION, "nginx version")
	nginxConfigurePath := flag.String("c", "", "configuration file for building nginx")
	modulesConfPath := flag.String("m", "", "configuration file for 3rd party modules")
	workParentDir := flag.String("d", "", "working directory")
	jobs := flag.Int("j", runtime.NumCPU(), "jobs to build nginx")
	verbose := flag.Bool("verbose", false, "verbose mode")
	pcreStatic := flag.Bool("pcre", false, "embedded PCRE statically")
	pcreVersion := flag.String("pcreversion", PCRE_VERSION, "PCRE version")
	openSSLStatic := flag.Bool("openssl", false, "embedded PCRE statically")
	openSSLVersion := flag.String("opensslversion", OPENSSL_VERSION, "OpenSSL version")
	zlibStatic := flag.Bool("zlib", false, "embedded zlib statically")
	zlibVersion := flag.String("zlibversion", ZLIB_VERSION, "zlib version")
	clear := flag.Bool("clear", false, "remove entries in working directory")
	versionPrint := flag.Bool("version", false, "print nginx-build versions")
	versionsPrint := flag.Bool("versions", false, "print nginx versions")
	openResty := flag.Bool("openresty", false, "download openresty instead of nginx")
	openRestyVersion := flag.String("openrestyversion", OPENRESTY_VERSION, "openresty version")
	flag.Parse()

	if *versionPrint {
		printNginxBuildVersion()
		os.Exit(0)
	}

	if *versionsPrint {
		printNginxVersions()
		os.Exit(0)
	}

	printFirstMsg()

	// set verbose mode
	VerboseEnabled = *verbose

	var nginxBuilder Builder
	if *openResty {
		nginxBuilder = makeBuilder(COMPONENT_OPENRESTY, *openRestyVersion)
	} else {
		nginxBuilder = makeBuilder(COMPONENT_NGINX, *version)
	}
	pcreBuilder := makeBuilder(COMPONENT_PCRE, *pcreVersion)
	openSSLBuilder := makeBuilder(COMPONENT_OPENSSL, *openSSLVersion)
	zlibBuilder := makeBuilder(COMPONENT_ZLIB, *zlibVersion)

	// change default umask
	_ = syscall.Umask(0)

	versionCheck(*version)

	nginxConfigure, err := fileGetContents(*nginxConfigurePath)
	if err != nil {
		log.Fatal(err.Error())
	}
	nginxConfigure = normalizeConfigure(nginxConfigure)

	modules3rd, err := loadModules3rdFile(*modulesConfPath)
	if err != nil {
		log.Fatal(err.Error())
	}

	if len(*workParentDir) == 0 {
		log.Fatal("set working directory with -d")
	}

	if !fileExists(*workParentDir) {
		err := os.Mkdir(*workParentDir, 0755)
		if err != nil {
			log.Fatalf("Failed to create working directory(%s) does not exist.", *workParentDir)
		}
	}

	var workDir string
	if *openResty {
		workDir = *workParentDir + "/" + *openRestyVersion
	} else {
		workDir = *workParentDir + "/" + *version
	}
	if *clear {
		err := clearWorkDir(workDir)
		if err != nil {
			log.Fatal(err.Error())
		}
	}

	if !fileExists(workDir) {
		err := os.Mkdir(workDir, 0755)
		if err != nil {
			log.Fatalf("Failed to create working directory(%s) does not exist.", workDir)
		}
	}

	rootDir := saveCurrentDir()
	// cd workDir
	err = os.Chdir(workDir)
	if err != nil {
		log.Fatal(err.Error())
	}

	if *pcreStatic {
		parallels++
		go downloadAndExtractParallel(&pcreBuilder, done)
	}

	if *openSSLStatic {
		parallels++
		go downloadAndExtractParallel(&openSSLBuilder, done)
	}

	if *zlibStatic {
		parallels++
		go downloadAndExtractParallel(&zlibBuilder, done)
	}

	parallels++
	go downloadAndExtractParallel(&nginxBuilder, done)

	if len(modules3rd) > 0 {
		parallels += len(modules3rd)
		for _, m := range modules3rd {
			go downloadAndExtractModule3rdParallel(m, done)
		}

	}

	// wait until all downloading processes by goroutine finish
	for i := 0; i < parallels; i++ {
		<-done
	}

	if len(modules3rd) > 0 {
		for _, m := range modules3rd {
			provideModule3rd(&m)
		}
	}

	// cd workDir/nginx-${version}
	os.Chdir(nginxBuilder.sourcePath())

	if *pcreStatic {
		dependencies = append(dependencies, makeStaticLibrary(&pcreBuilder))
	}

	if *openSSLStatic {
		dependencies = append(dependencies, makeStaticLibrary(&openSSLBuilder))
	}

	if *zlibStatic {
		dependencies = append(dependencies, makeStaticLibrary(&zlibBuilder))
	}

	log.Printf("Generate configure script for %s.....", nginxBuilder.sourcePath())

	if *pcreStatic && pcreBuilder.isIncludeWithOption(nginxConfigure) {
		log.Printf(pcreBuilder.warnMsgWithLibrary())
	}

	if *openSSLStatic && openSSLBuilder.isIncludeWithOption(nginxConfigure) {
		log.Printf(openSSLBuilder.warnMsgWithLibrary())

	}

	if *zlibStatic && zlibBuilder.isIncludeWithOption(nginxConfigure) {
		log.Printf(zlibBuilder.warnMsgWithLibrary())
	}

	if strings.Contains(nginxConfigure, "--add-module=") {
		log.Println("[warn]Using '--add-module' is discouraged. Instead give ini-file with '-m' to 'nginx-build'")
	}

	configureScript := configureGen(nginxConfigure, modules3rd, dependencies)
	err = ioutil.WriteFile("./nginx-configure", []byte(configureScript), 0655)
	if err != nil {
		log.Fatalf("Failed to generate configure script for %s", nginxBuilder.sourcePath())
	}
	log.Printf("Configure %s.....", nginxBuilder.sourcePath())
	err = configureNginx()
	if err != nil {
		log.Fatalf("Failed to configure %s", nginxBuilder.sourcePath())
	}

	log.Printf("Build %s.....", nginxBuilder.sourcePath())
	if *openSSLStatic {
		// This is a workaround for protecting a failure of building nginx with OpenSSL.
		// Unfortunately build of nginx with static OpenSSL fails by multi-CPUs.
		*jobs = 1
	}
	err = buildNginx(*jobs)
	if err != nil {
		fmt.Println(err.Error())
		log.Fatalf("Failed to build %s", nginxBuilder.sourcePath())
	}

	printLastMsg(workDir, nginxBuilder.sourcePath(), *openResty)

	// cd rootDir
	os.Chdir(rootDir)
}
